File: src/com/jetbrains/php/tools/quality/phpstan/PhpStanOptionsPanel.java
package com.jetbrains.php.tools.quality.phpstan;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.DocumentAdapter;
import com.intellij.ui.JBIntSpinner;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBTextField;
import com.jetbrains.php.config.interpreters.PhpInterpreter;
import com.jetbrains.php.config.interpreters.PhpTextFieldWithSdkBasedBrowse;
import com.jetbrains.php.tools.quality.QualityToolConfigurationComboBox;
import com.jetbrains.php.tools.quality.QualityToolType;
import com.jetbrains.php.tools.quality.QualityToolsOptionsPanel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.DocumentEvent;
import java.net.URL;

import static com.intellij.openapi.vfs.VfsUtil.findFileByURL;
import static com.intellij.openapi.vfs.VfsUtilCore.convertToURL;
import static com.intellij.openapi.vfs.VfsUtilCore.pathToUrl;

public class PhpStanOptionsPanel extends QualityToolsOptionsPanel {
  private JPanel myOptionsPanel;
  private JBCheckBox myFullProjectRunJBCheckBox;
  private JBTextField myMemoryLimitTextField;
  private JBIntSpinner myJBIntSpinner;
  private PhpTextFieldWithSdkBasedBrowse myConfigPathTextField;
  private PhpTextFieldWithSdkBasedBrowse myAutoloadPathTextField;
  private final QualityToolConfigurationComboBox myComboBox;

  public PhpStanOptionsPanel(Project project,
                             QualityToolConfigurationComboBox comboBox,
                             Runnable validate) {
    super(project, validate, PhpStanQualityToolType.INSTANCE);
    myComboBox = comboBox;
    PhpStanProjectConfiguration configuration = PhpStanProjectConfiguration.getInstance(project);
    myFullProjectRunJBCheckBox.setSelected(configuration.isFullProject());
    myMemoryLimitTextField.setText(configuration.getMemoryLimit());
    myJBIntSpinner.setNumber(configuration.getLevel());
    myConfigPathTextField.setText(configuration.getConfig());
    myConfigPathTextField
      .init(project, getSdkAdditionalData(project, comboBox), PhpStanBundle.message("phpstan.configuration.file"), true, false);
    myAutoloadPathTextField.setText(configuration.getAutoload());
    myAutoloadPathTextField
      .init(project, getSdkAdditionalData(project, comboBox), PhpStanBundle.message("phpstan.autoload.file"), true, false);
    myConfigPathTextField.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
      @Override
      protected void textChanged(@NotNull DocumentEvent e) {
        validate.run();
      }
    });
  }

  private void createUIComponents() {
    myJBIntSpinner = new JBIntSpinner(4, 0, 8);
  }

  @Override
  public JPanel getOptionsPanel() {
    return myOptionsPanel;
  }

  @Override
  public void reset() {
    PhpStanProjectConfiguration configuration = PhpStanProjectConfiguration.getInstance(myProject);
    myFullProjectRunJBCheckBox.setSelected(configuration.isFullProject());
    myMemoryLimitTextField.setText(configuration.getMemoryLimit());
    myJBIntSpinner.setNumber(configuration.getLevel());
    myConfigPathTextField.setText(configuration.getConfig());
    myAutoloadPathTextField.setText(configuration.getAutoload());
  }

  @Override
  public boolean isModified() {
    PhpStanProjectConfiguration configuration = PhpStanProjectConfiguration.getInstance(myProject);
    if (myFullProjectRunJBCheckBox.isSelected() != configuration.isFullProject()) return true;
    if (!StringUtil.equals(myMemoryLimitTextField.getText(), configuration.getMemoryLimit())) return true;
    if (myJBIntSpinner.getNumber() != configuration.getLevel()) return true;
    if (!StringUtil.equals(myConfigPathTextField.getText(), configuration.getConfig())) return true;
    if (!StringUtil.equals(myAutoloadPathTextField.getText(), configuration.getAutoload())) return true;
    return false;
  }

  @Override
  public void apply() {
    PhpStanProjectConfiguration configuration = PhpStanProjectConfiguration.getInstance(myProject);
    configuration.setFullProject(myFullProjectRunJBCheckBox.isSelected());
    configuration.setMemoryLimit(myMemoryLimitTextField.getText());
    configuration.setLevel(myJBIntSpinner.getNumber());
    configuration.setConfig(myConfigPathTextField.getText());
    configuration.setAutoload(myAutoloadPathTextField.getText());
  }

  @Override
  protected @Nullable String validatePath() {
    PhpInterpreter interpreter = getSelectedInterpreter(myProject, myComboBox);
    if (interpreter != null && interpreter.isRemote()) {
      //TODO: validate remote path?
      return null;
    }
    final URL url = convertToURL(pathToUrl(myConfigPathTextField.getText()));
    if (url == null || findFileByURL(url) == null) {
      return PhpStanBundle.message("config.file.doesnt.exist");
    }
    return null;
  }
}

File: src/com/jetbrains/php/tools/quality/phpstan/PhpStanOptionsPanel.kt
package com.jetbrains.php.tools.quality.phpstan

import com.intellij.openapi.project.Project
import com.intellij.openapi.util.text.StringUtil
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VfsUtilCore
import com.intellij.ui.DocumentAdapter
import com.intellij.ui.JBIntSpinner
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBTextField
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.panel
import com.jetbrains.php.config.interpreters.PhpTextFieldWithSdkBasedBrowse
import com.jetbrains.php.tools.quality.QualityToolConfigurationComboBox
import com.jetbrains.php.tools.quality.QualityToolsOptionsPanel
import com.jetbrains.php.tools.quality.ui.QualityToolRateLimitPanel
import com.jetbrains.php.tools.quality.ui.QualityToolRateLimitUI
import javax.swing.JPanel
import javax.swing.event.DocumentEvent

class PhpStanOptionsPanel(
  project: Project,
  private val myComboBox: QualityToolConfigurationComboBox<*>,
  validate: Runnable,
) : QualityToolsOptionsPanel(project, validate, PhpStanQualityToolType.INSTANCE) {

  private val myJBIntSpinner = JBIntSpinner(4, 0, 8)
  private val myMemoryLimitTextField = JBTextField()
  private val myConfigPathTextField = PhpTextFieldWithSdkBasedBrowse()
  private val myAutoloadPathTextField = PhpTextFieldWithSdkBasedBrowse()
  private val myRateLimitPanel = QualityToolRateLimitPanel()
  private lateinit var myFullProjectRunJBCheckBox: JBCheckBox
  private lateinit var myEditorModeJBCheckBox: JBCheckBox

  private val myOptionsPanel = panel {
    row {
      myFullProjectRunJBCheckBox = checkBox(PhpStanBundle.message("phpstan.checkbox.full.project.run")).component
    }
    row {
      myEditorModeJBCheckBox = checkBox(PhpStanBundle.message("phpstan.checkbox.editor.mode")).component
    }
    row(PhpStanBundle.message("label.level")) {
      cell(myJBIntSpinner)
    }
    row(PhpStanBundle.message("label.configuration.file")) {
      cell(myConfigPathTextField).align(AlignX.FILL)
    }
    row(PhpStanBundle.message("label.autoload.file")) {
      cell(myAutoloadPathTextField).align(AlignX.FILL)
    }
    row(PhpStanBundle.message("phpstan.label.options")) {
      cell(myMemoryLimitTextField).align(AlignX.FILL)
    }
    row {
      cell(myRateLimitPanel).align(AlignX.FILL)
    }
  }

  init {
    val configuration = PhpStanOptionsConfiguration.getInstance(project)
    myFullProjectRunJBCheckBox.isSelected = configuration.isFullProject
    myEditorModeJBCheckBox.isSelected = configuration.isEditorMode
    myMemoryLimitTextField.text = configuration.memoryLimit
    myJBIntSpinner.number = configuration.level
    myRateLimitPanel.configure(QualityToolRateLimitUI.DEFAULT_UI)
    myRateLimitPanel.reset(configuration.rateLimitSettings)
    myConfigPathTextField.text = configuration.config
    myConfigPathTextField.init(project, getSdkAdditionalData(project, myComboBox),
                               PhpStanBundle.message("phpstan.configuration.file"), true, false)
    myAutoloadPathTextField.text = configuration.autoload
    myAutoloadPathTextField.init(project, getSdkAdditionalData(project, myComboBox),
                                 PhpStanBundle.message("phpstan.autoload.file"), true, false)
    myConfigPathTextField.textField.document.addDocumentListener(object : DocumentAdapter() {
      override fun textChanged(e: DocumentEvent) {
        validate.run()
      }
    })
  }

  override fun getOptionsPanel(): JPanel = myOptionsPanel

  override fun reset() {
    val configuration = PhpStanOptionsConfiguration.getInstance(myProject)
    myFullProjectRunJBCheckBox.isSelected = configuration.isFullProject
    myEditorModeJBCheckBox.isSelected = configuration.isEditorMode
    myMemoryLimitTextField.text = configuration.memoryLimit
    myJBIntSpinner.number = configuration.level
    myRateLimitPanel.reset(configuration.rateLimitSettings)
    myConfigPathTextField.text = configuration.config
    myAutoloadPathTextField.text = configuration.autoload
  }

  override fun isModified(): Boolean {
    val configuration = PhpStanOptionsConfiguration.getInstance(myProject)
    if (myFullProjectRunJBCheckBox.isSelected != configuration.isFullProject) return true
    if (myEditorModeJBCheckBox.isSelected != configuration.isEditorMode) return true
    if (!StringUtil.equals(myMemoryLimitTextField.text, configuration.memoryLimit)) return true
    if (myJBIntSpinner.number != configuration.level) return true
    if (myRateLimitPanel.isModified(configuration.rateLimitSettings)) return true
    if (!StringUtil.equals(myConfigPathTextField.text, configuration.config)) return true
    if (!StringUtil.equals(myAutoloadPathTextField.text, configuration.autoload)) return true
    return false
  }

  override fun apply() {
    val configuration = PhpStanOptionsConfiguration.getInstance(myProject)
    configuration.isFullProject = myFullProjectRunJBCheckBox.isSelected
    configuration.isEditorMode = myEditorModeJBCheckBox.isSelected
    configuration.memoryLimit = myMemoryLimitTextField.text
    configuration.level = myJBIntSpinner.number
    myRateLimitPanel.applyTo(configuration.rateLimitSettings)
    configuration.config = myConfigPathTextField.text
    configuration.autoload = myAutoloadPathTextField.text
  }

  override fun validatePath(): String? {
    val interpreter = getSelectedInterpreter(myProject, myComboBox)
    if (interpreter != null && interpreter.isRemote) {
      //TODO: validate remote path?
      return null
    }
    val url = VfsUtilCore.convertToURL(VfsUtilCore.pathToUrl(myConfigPathTextField.text))
    if (url == null || VfsUtil.findFileByURL(url) == null) {
      return PhpStanBundle.message("config.file.doesnt.exist")
    }
    return null
  }
}
